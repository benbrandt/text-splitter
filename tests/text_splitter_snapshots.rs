use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
