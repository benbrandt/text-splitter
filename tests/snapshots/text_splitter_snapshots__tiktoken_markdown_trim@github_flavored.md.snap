---
source: tests/text_splitter_snapshots.rs
expression: chunks
input_file: tests/inputs/markdown/github_flavored.md
---
- "# Headers"
- "```\n# h1 Heading 8-)\n## h2 Heading"
- "### h3 Heading\n#### h4 Heading\n##### h5 Heading"
- "###### h6 Heading"
- "Alternatively, for H1 and H2, an underline-ish style:"
- "Alt-H1\n======\n\nAlt-H2\n------\n```"
- "# h1 Heading 8-)"
- "## h2 Heading"
- "### h3 Heading"
- "#### h4 Heading"
- "##### h5 Heading"
- "###### h6 Heading"
- "Alternatively, for H1 and H2, an underline-ish style:"
- "Alt-H1\n======\n\nAlt-H2\n------\n\n------"
- "# Emphasis"
- "```"
- "Emphasis, aka italics, with *asterisks* or"
- _underscores_.
- "Strong emphasis, aka bold, with **asterisks** or __underscores__."
- Combined emphasis with **asterisks and _underscores_**.
- Strikethrough uses two tildes. ~~Scratch this.~~
- "**This is bold text**\n\n__This is bold text__"
- "*This is italic text*\n\n_This is italic text_"
- "~~Strikethrough~~\n```"
- "Emphasis, aka italics, with *asterisks* or"
- _underscores_.
- "Strong emphasis, aka bold, with **asterisks** or __underscores__."
- Combined emphasis with **asterisks and _underscores_**.
- Strikethrough uses two tildes. ~~Scratch this.~~
- "**This is bold text**\n\n__This is bold text__"
- "*This is italic text*\n\n_This is italic text_"
- "~~Strikethrough~~\n\n------"
- "# Lists"
- "```\n1. First ordered list item\n2. Another item"
- "⋅⋅* Unordered sub-list.\n1."
- "Actual numbers don't matter, just that it's a number"
- "⋅⋅1. Ordered sub-list\n4."
- And another item.
- ⋅⋅⋅You can have properly indented paragraphs
- within list items.
- "Notice the blank line above, and the leading spaces (at least one, but"
- "we'll use three here to also align the raw Markdown)."
- ⋅⋅⋅To have a line break without a
- "paragraph, you will need to use two trailing spaces."
- ⋅⋅
- "⋅⋅⋅Note that this line is separate,"
- but within the same paragraph.⋅⋅
- ⋅⋅⋅(This is contrary to the typical
- "GFM line break behaviour, where trailing spaces are not required.)"
- "* Unordered list can use asterisks\n- Or minuses"
- "+ Or pluses\n\n1. Make my changes\n    1."
- "Fix bug\n    2. Improve formatting\n        - Make the headings bigger"
- "2. Push my commits to GitHub\n3. Open a pull request"
- "    * Describe my changes\n    * Mention all the members of my team"
- "* Ask for feedback"
- "+ Create a list by starting a line with `+`, `-`, or `"
- "*`\n+ Sub-lists are made by indenting 2 spaces:"
- "- Marker character change forces new list start:"
- "* Ac tristique libero volutpat at"
- + Facilisis in pretium nisl aliquet
- "    - Nulla volutpat aliquam velit\n+ Very easy!\n```"
- 1. First ordered list item
- "2. Another item\n⋅⋅*"
- Unordered sub-list.
- "1. Actual numbers don't matter, just that it's a number"
- ⋅⋅1. Ordered sub-list
- 4. And another item.
- ⋅⋅⋅You can have properly indented paragraphs
- within list items.
- "Notice the blank line above, and the leading spaces (at least one, but"
- "we'll use three here to also align the raw Markdown)."
- ⋅⋅⋅To have a line break without a
- "paragraph, you will need to use two trailing spaces."
- ⋅⋅
- "⋅⋅⋅Note that this line is separate,"
- but within the same paragraph.⋅⋅
- ⋅⋅⋅(This is contrary to the typical
- "GFM line break behaviour, where trailing spaces are not required.)"
- "* Unordered list can use asterisks\n- Or minuses"
- + Or pluses
- "1. Make my changes\n    1. Fix bug"
- "    2. Improve formatting\n        - Make the headings bigger"
- "2. Push my commits to GitHub\n3. Open a pull request"
- " * Describe my changes\n    * Mention all the members of my team"
- "* Ask for feedback"
- "+ Create a list by starting a line with `+`, `-`, or `"
- "*`\n+ Sub-lists are made by indenting 2 spaces:"
- "- Marker character change forces new list start:"
- "* Ac tristique libero volutpat at"
- + Facilisis in pretium nisl aliquet
- "- Nulla volutpat aliquam velit\n+ Very easy!\n\n------"
- "# Task lists"
- "```\n- [x] Finish my changes"
- "- [ ] Push my commits to GitHub\n- [ ] Open a pull request"
- "- [x] @mentions, #refs, [links](), **"
- "formatting**, and <del>tags</del> supported"
- "- [x] list syntax required (any unordered or ordered list supported)"
- "- [x] this is a complete item"
- "- [ ] this is an incomplete item\n```"
- "- [x] Finish my changes\n- [ ] Push my commits to GitHub"
- "- [ ] Open a pull request"
- "- [x] @mentions, #refs, [links](), **"
- "formatting**, and <del>tags</del> supported"
- "- [x] list syntax required (any unordered or ordered list supported)"
- "- [ ] this is a complete item"
- "- [ ] this is an incomplete item\n\n------"
- "# Ignoring Markdown formatting"
- "You can tell GitHub to ignore (or escape) Markdown formatting by using \\ before"
- the Markdown character.
- "```"
- "Let's rename \\*our-new-project\\* to \\*our-old-project"
- "\\*.\n```"
- "Let's rename \\*our-new-project\\* to \\*our-old-project"
- "\\*.\n\n------"
- "# Links"
- "```\n[I'm an inline-style link](https://www.google.com)"
- "[I'm an inline-style link with title](https://www.google.com \""
- "Google's Homepage\")"
- "[I'm a reference-style link][Arbitrary case-insensitive reference text]"
- "[I'm a relative reference to a repository file](../blob/master/LICENSE)"
- "[You can use numbers for reference-style link definitions][1]"
- "Or leave it empty and use the [link text itself]."
- URLs and URLs in angle brackets will automatically get turned into links.
- "http://www.example.com or <http://www.example.com> and sometimes"
- "example.com (but not on Github, for example)."
- Some text to show that the reference links can follow later.
- "[arbitrary case-insensitive reference text]: https://www.mozilla.org"
- "[1]: http://slashdot.org"
- "[link text itself]: http://www.reddit.com\n```"
- "[I'm an inline-style link](https://www.google.com)"
- "[I'm an inline-style link with title](https://www.google.com \""
- "Google's Homepage\")"
- "[I'm a reference-style link][Arbitrary case-insensitive reference text]"
- "[I'm a relative reference to a repository file](../blob/master/LICENSE)"
- "[You can use numbers for reference-style link definitions][1]"
- "Or leave it empty and use the [link text itself]."
- URLs and URLs in angle brackets will automatically get turned into links.
- "http://www.example.com or <http://www.example.com>"
- " and sometimes\nexample.com (but not on Github, for example)."
- Some text to show that the reference links can follow later.
- "[arbitrary case-insensitive reference text]: https://www.mozilla.org"
- "[1]: http://slashdot.org"
- "[link text itself]: http://www.reddit.com\n\n------"
- "# Images"
- "```\nHere's our logo (hover to see the title text):"
- "Inline-style:\n!["
- "alt text](https://github.com/adam-p/markdown-here/raw/master"
- "/src/common/images/icon48.png \"Logo Title Text 1\")\n\nReference-style:"
- "![alt text][logo]"
- "[logo]: https://github.com/adam-p/markdown-here/raw/master"
- "/src/common/images/icon48.png \"Logo Title Text 2\"\n\n!["
- "Minion](https://octodex.github.com/images/minion.png)\n!["
- "Stormtroopocat](https://octodex.github.com/images/"
- "stormtroopocat.jpg \"The Stormtroopocat\")"
- "Like links, Images also have a footnote style syntax\n\n!["
- "Alt text][id]"
- "With a reference later in the document defining the URL location:"
- "[id]: https://octodex.github.com/images/dojocat.jpg"
- "\"The Dojocat\"\n```"
- "Here's our logo (hover to see the title text):"
- "Inline-style:"
- "!["
- "alt text](https://github.com/adam-p/markdown-here/raw/master"
- "/src/common/images/icon48.png \"Logo Title Text 1\")"
- "Reference-style:\n![alt text][logo]"
- "[logo]: https://github.com/adam-p/markdown-here/raw/master"
- "/src/common/images/icon48.png \"Logo Title Text 2\""
- "![Minion](https://octodex.github.com/images/minion.png)"
- "!["
- "Stormtroopocat](https://octodex.github.com/images/"
- "stormtroopocat.jpg \"The Stormtroopocat\")"
- "Like links, Images also have a footnote style syntax"
- "![Alt text][id]"
- "With a reference later in the document defining the URL location:"
- "[id]: https://octodex.github.com/images/dojocat.jpg"
- "\"The Dojocat\"\n\n------"
- "# [Footnotes](https://github.com/markdown-it/markdown-it-"
- footnote)
- "```\nFootnote 1 link[^first]."
- "Footnote 2 link[^second]."
- "Inline footnote^[Text of inline footnote] definition."
- "Duplicated footnote reference[^second]."
- "[^first]: Footnote **can have markup**\n\n    and multiple paragraphs."
- "[^second]: Footnote text.\n```\n\nFootnote 1 link"
- "[^first].\n\nFootnote 2 link[^second]."
- "Inline footnote^[Text of inline footnote] definition."
- "Duplicated footnote reference[^second]."
- "[^first]: Footnote **can have markup**\n\n    and multiple paragraphs."
- "[^second]: Footnote text.\n\n------"
- "# Code and Syntax Highlighting"
- "```\nInline `code` has `back-ticks around` it.\n```"
- "Inline `code` has `back-ticks around` it."
- "```c#\nusing System.IO.Compression;"
- "#pragma warning disable 414, 3021\n\nnamespace MyApplication\n{"
- "    [Obsolete(\"...\")]\n    class Program : IInterface\n    {"
- "        public static List<int> JustDoIt(int count)\n        {"
- "Console.WriteLine($\"Hello {Name}!\");"
- "return new List<int>(new int[] { 1, 2,"
- "3 })\n        }\n    }\n}\n```"
- "```css\n@font-face {"
- "font-family: Chunkfive; src: url('Chunkfive.otf');"
- "}\n\nbody, .usertext {"
- "color: #F0F0F0; background: #600;"
- "  font-family: Chunkfive, sans;\n}\n\n@import url(print.css);"
- "@media print {\n  a[href^=http]::after {"
- "    content: attr(href)\n  }\n}\n```"
- "```javascript\nfunction $initHighlight(block, cls) {\n  try {"
- "if (cls.search(/\\bno\\-highlight\\b/) != -1)"
- "return process(block, true, 0x0F) +"
- "             ` class=\"${cls}\"`;\n  } catch (e) {"
- "    /* handle exception */\n  }"
- for (var i = 0 / 2; i < classes.length;
- "i++) {\n    if (checkCondition(classes[i]) === undefined)"
- "      console.log('undefined');\n  }\n}\n\nexport  $initHighlight;\n```"
- "```php\nrequire_once 'Zend/Uri/Http.php';"
- "namespace Location\\Web;\n\ninterface Factory\n{\n    static function _factory();\n}"
- "abstract class URI extends BaseURI implements Factory\n{\n    abstract function test();"
- public static $st1 = 1;
- "    const ME = \"Yo\";\n    var $list = NULL;"
- "    private $var;\n\n    /**\n     * Returns a URI\n     *"
- "     * @return URI\n     */"
- "static public function _factory($stats = array(), $uri = 'http')"
- "    {\n        echo __METHOD__;"
- "$uri = explode(':', $uri, 0b10);"
- "$schemeSpecific = isset($uri[1]) ?"
- "$uri[1] : '';\n        $desc = 'Multi"
- "line description';\n\n        // Security check\n        if (!"
- "ctype_alnum($scheme)) {"
- "            throw new Zend_Uri_Exception('Illegal scheme');\n        }"
- "$this->var = 0 - self::$st;"
- "$this->list = list(Array(\"1\"=> 2, 2=>self"
- "::ME, 3 => \\Location\\Web\\URI::class));"
- "        return [\n            'uri'   => $uri,"
- "            'value' => null,\n        ];\n    }\n}"
- "echo URI::ME . URI::$st1;"
- "__halt_compiler () ; datahere\ndatahere\ndatahere */\ndatahere"
- "```\n\n------"
- "# Tables"
- "```\nColons can be used to align columns."
- "| Tables        | Are           | Cool  |"
- "| ------------- |:-------------:| -----:|"
- "| col 3 is      | right-aligned | $1600 |"
- "| col 2 is      | centered      |   $12 |"
- "| zebra stripes | are neat      |    $1 |"
- There must be at least 3 dashes separating each header cell.
- "The outer pipes (|) are optional, and you don't need to make"
- "the\nraw Markdown line up prettily."
- "You can also use inline Markdown.\n\nMarkdown | Less | Pretty"
- "--- | --- | ---"
- "*Still* | `renders` | **nicely**"
- "1 | 2 | 3\n\n| First Header  | Second Header |"
- "| ------------- | ------------- |\n| Content Cell  | Content Cell  |"
- "| Content Cell  | Content Cell  |\n\n| Command | Description |"
- "| --- | --- |\n| git status | List all new or modified files |"
- "| git diff | Show file differences that haven't been staged |"
- "| Command | Description |\n| --- | --- |"
- "| `git status` | List all *new or modified* files |"
- "| `git diff` | Show file differences that **haven't been** staged"
- "|\n\n| Left-aligned | Center-aligned | Right-aligned |"
- "| :---         |     :---:      |          ---: |"
- "| git status   | git status     | git status    |"
- "| git diff     | git diff       | git diff      |"
- "| Name     | Character |\n| ---      | ---       |"
- "| Backtick | `         |\n| Pipe     | \\|        |\n```"
- Colons can be used to align columns.
- "| Tables        | Are           | Cool  |"
- "| ------------- |:-------------:| -----:|"
- "| col 3 is      | right-aligned | $1600 |"
- "| col 2 is      | centered      |   $12 |"
- "| zebra stripes | are neat      |    $1 |"
- There must be at least 3 dashes separating each header cell.
- "The outer pipes (|) are optional, and you don't need to make"
- the
- raw Markdown line up prettily. You can also use inline Markdown.
- "Markdown | Less | Pretty\n--- | --- | ---"
- "*Still* | `renders` | **nicely**"
- 1 | 2 | 3
- "| First Header  | Second Header |\n| ------------- | ------------- |"
- "| Content Cell  | Content Cell  |"
- "| Content Cell  | Content Cell  |"
- "| Command | Description |\n| --- | --- |"
- "| git status | List all new or modified files |"
- "| git diff | Show file differences that haven't been staged |"
- "| Command | Description |\n| --- | --- |"
- "| `git status` | List all *new or modified* files |"
- "| `git diff` | Show file differences that **haven't been** staged"
- "|"
- "| Left-aligned | Center-aligned | Right-aligned |"
- "| :---         |     :---:      |          ---: |"
- "| git status   | git status     | git status    |"
- "| git diff     | git diff       | git diff      |"
- "| Name     | Character |\n| ---      | ---       |"
- "| Backtick | `         |\n| Pipe     | \\|        |\n\n------"
- "# Blockquotes"
- "```\n> Blockquotes are very handy in email to emulate reply text."
- "> This line is part of the same quote.\n\nQuote break."
- "> This is a very long line that will still be quoted properly when it wraps"
- "."
- "Oh boy let's keep writing to make sure this is long enough to actually wrap"
- for everyone.
- "Oh, you can *put* **Markdown** into a blockquote."
- "> Blockquotes can also be nested..."
- ">> ...by using additional greater-than signs right next to each other..."
- "> > > ...or with spaces between arrows.\n```"
- "> Blockquotes are very handy in email to emulate reply text.\n>"
- "This line is part of the same quote.\n\nQuote break."
- "> This is a very long line that will still be quoted properly when it wraps"
- ". Oh boy let'"
- s keep writing to make sure this is long enough to actually wrap for everyone.
- "Oh, you can *put* **Markdown** into a blockquote."
- "> Blockquotes can also be nested...\n>"
- "> ...by using additional greater-than signs right next to each other...\n> >"
- "> ...or with spaces between arrows.\n\n------"
- "# Inline HTML"
- "```"
- "<dl>\n  <dt>Definition list</dt>"
- "<dd>Is something people use sometimes.</dd>"
- "<dt>Markdown in HTML</dt>"
- "<dd>Does *not* work **very** well."
- "Use HTML <em>tags</em>.</dd>\n</dl>\n```"
- "<dl>\n  <dt>Definition list</dt>"
- "<dd>Is something people use sometimes.</dd>"
- "<dt>Markdown in HTML</dt>"
- "<dd>Does *not* work **very** well."
- "Use HTML <em>tags</em>.</dd>\n</dl>\n\n------"
- "# Horizontal Rules"
- "```"
- "Three or more...\n\n---\n\nHyphens\n\n***\n\nAsterisks\n\n___"
- "Underscores\n```\n\nThree or more...\n\n---\n\nHyphens\n\n***"
- "Asterisks\n\n___\n\nUnderscores\n\n------"
- "# YouTube Videos"
- "```"
- "<a href=\"http://www.youtube.com/watch?"
- "feature=player_embedded&v=YOUTUBE_VIDEO_ID_HERE\" target=\""
- "_blank\">"
- "<img src=\"http://img.youtube.com/vi/YOUTUBE_VIDEO_ID_HERE"
- "/0.jpg\" alt=\"IMAGE ALT TEXT HERE\" width=\"240\" height"
- "=\"180\" border=\"10\">\n</a>\n```"
- "<a href=\"http://www.youtube.com/watch?"
- "feature=player_embedded&v=YOUTUBE_VIDEO_ID_HERE\" target=\""
- "_blank\">"
- "<img src=\"http://img.youtube.com/vi/YOUTUBE_VIDEO_ID_HERE"
- "/0.jpg\" alt=\"IMAGE ALT TEXT HERE\" width=\"240\" height"
- "=\"180\" border=\"10\">\n</a>"
- "```"
- "[!["
- "IMAGE ALT TEXT HERE](http://img.youtube.com/vi/"
- "YOUTUBE_VIDEO_ID_HERE/0.jpg)](http://www.youtube.com"
- "/watch?v=YOUTUBE_VIDEO_ID_HERE)\n```"
- "[![IMAGE ALT TEXT HERE"
- "](https://upload.wikimedia.org/wikipedia/commons/thumb/e/ef/"
- YouTube_logo_2015.svg/1200px-YouTube_logo_2015
- ".svg.png)](https://www.youtube.com/watch?"
- v=ciawICBvQoE)
